File: src/lib/nodes/identifier.ts
import { ChatGroq } from "@langchain/groq";
import { AgentState } from "../types";
import { safeParseLlmJson, extractTextContent } from "../utils/parseJson";

export const identifierNode = async (state: AgentState): Promise<Partial<AgentState>> => {
  const companyInput = state.companyInput;
  if (!companyInput) {
    return {
      error: "No companyInput provided in state.",
      streamLog: ["⊗ ABORT — No input provided."],
    };
  }

  try {
    const llm = new ChatGroq({
      model: "llama-3.3-70b-versatile",
      temperature: 0,
      maxTokens: 1024,
    });

    const systemPrompt =
      "You are a financial data resolver. Return ONLY valid JSON, no markdown, no explanation.";
    const userPrompt = `Given this company name: ${companyInput}
Return this exact JSON:
{
  "resolvedName": "string",
  "ticker": "string | null",
  "exchange": "string | null",
  "country": "string",
  "sector": "string",
  "industry": "string",
  "isPublic": boolean,
  "founded": "string | null",
  "description": "string"
}
If company not found return: { "error": "Not found" }
ticker and exchange are null for private companies.`;

    const response = await llm.invoke([
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ]);

    const raw = extractTextContent(response.content);
    const { data: parsedResult, error } = safeParseLlmJson(raw, "IdentifierNode");

    if (error || !parsedResult) {
      return {
        error: `Failed to parse company data: ${error}`,
        streamLog: [`⊗ PARSE FAULT — ${error}`],
      };
    }

    const result = parsedResult as Record<string, unknown>;

    if (result.error) {
      return {
        error: String(result.error),
        streamLog: [`⊗ IDENTIFIER FAULT — ${result.error}`],
      };
    }

    // Rename resolvedName → name to match AgentState CompanyInfo type
    const { resolvedName, ...rest } = result;
    const companyInfo = { name: resolvedName, ...rest };

    return {
      companyInfo: companyInfo as AgentState["companyInfo"],
      streamLog: [`◈ RESOLVED: ${resolvedName} (${(companyInfo as any).ticker ?? "Private"})`],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      error: `Failed to resolve company data: ${errorMessage}`,
      streamLog: [`⊗ IDENTIFIER FAULT — ${errorMessage}`],
    };
  }
};

File: backend/nodes/identifier.py
"""
nodes/identifier.py — Company Identifier Node
Port of src/lib/nodes/identifier.ts
"""

import logging
from langchain_groq import ChatGroq
from langchain_core.messages import SystemMessage, HumanMessage

from state import AgentState
from utils.parse_json import safe_parse_llm_json

logger = logging.getLogger(__name__)


async def identifier_node(state: AgentState) -> dict:
    company_input = state.get("companyInput")
    if not company_input:
        return {
            "error": "No companyInput provided in state.",
            "streamLog": ["⊗ ABORT — No input provided."],
        }

    try:
        llm = ChatGroq(
            model="llama-3.3-70b-versatile",
            temperature=0,
            max_tokens=1024,
        )

        system_prompt = (
            "You are a financial data resolver. Return ONLY valid JSON, no markdown, no explanation."
        )
        user_prompt = f"""Given this company name: {company_input}
Return this exact JSON:
{{
  "resolvedName": "string",
  "ticker": "string | null",
  "exchange": "string | null",
  "country": "string",
  "sector": "string",
  "industry": "string",
  "isPublic": true or false,
  "founded": "string | null",
  "description": "string"
}}
If company not found return: {{ "error": "Not found" }}
ticker and exchange are null for private companies."""

        response = await llm.ainvoke([
            SystemMessage(content=system_prompt),
            HumanMessage(content=user_prompt),
        ])

        raw = response.content if isinstance(response.content, str) else str(response.content)
        data, error = safe_parse_llm_json(raw, "IdentifierNode")

        if error or data is None:
            return {
                "error": f"Failed to parse company data: {error}",
                "streamLog": [f"⊗ PARSE FAULT — {error}"],
            }

        if "error" in data:
            return {
                "error": str(data["error"]),
                "streamLog": [f"⊗ IDENTIFIER FAULT — {data['error']}"],
            }

        # Map resolvedName → name (matches the TS rename pattern)
        resolved_name = data.pop("resolvedName", None)
        company_info = {"name": resolved_name, **data}

        return {
            "companyInfo": company_info,
            "streamLog": [f"◈ RESOLVED: {resolved_name} ({company_info.get('ticker') or 'Private'})"],
        }

    except Exception as exc:
        error_message = str(exc)
        return {
            "error": f"Failed to resolve company data: {error_message}",
            "streamLog": [f"⊗ IDENTIFIER FAULT — {error_message}"],
        }
